stl/inc/algorithm

```diff
@@ -3567,6 +3567,77 @@ _FwdIt2 unique_copy(_ExPo&&, _FwdIt1 _First, _FwdIt1 _Last, _FwdIt2 _Dest) noexc
 }
 #endif // _HAS_CXX17
 
+#ifdef __cpp_lib_concepts
+namespace ranges {
+    // VARIABLE ranges::reverse
+    class _Reverse_fn : private _Not_quite_object {
+    public:
+        using _Not_quite_object::_Not_quite_object;
+
+        // clang-format off
+        template <bidirectional_iterator _It, sentinel_for<_It> _Se>
+            requires permutable<_It>
+        constexpr _It operator()(_It _First, _Se _Last) const {
+            _Adl_verify_range(_First, _Last);
+            auto _UFirst = _Get_unwrapped(_STD move(_First));
+            auto _ULast  = _Get_final_iterator_unwrapped<_It>(_UFirst, _STD move(_Last));
+            _Seek_wrapped(_First, _ULast);
+            _Reverse_unchecked(_STD move(_UFirst), _STD move(_ULast));
+            return _First;
+        }
+
+        template <bidirectional_range _Rng>
+            requires permutable<iterator_t<_Rng>>
+        constexpr borrowed_iterator_t<_Rng> operator()(_Rng&& _Range) const {
+            auto _ULast = _Get_final_iterator_unwrapped(_Range);
+            _Reverse_unchecked(_Ubegin(_Range), _ULast);
+            return _Rewrap_iterator(_Range, _STD move(_ULast));
+        }
+        // clang-format on
+    private:
+        template <class _It>
+        static constexpr void _Reverse_unchecked(_It _First, _It _Last) {

```

Comment:
Also given that we will need it for `ranges::rotate` could you directly make it a free function?

Reply:
This isn't a proper `_Reverse_unchecked` since it accepts iterator pairs unlike the algorithm which accepts iterator+sentinel pairs; I'll rename it to `_Reverse_common` (we often use "common" in Ranges as in `common_range`/`common_iterator` to names things which have or require the same iterator and sentinel type).

I can certainly promote the resulting function to namespace scope.

Reply:
Aside: we're going to need some custom logic in `rotate` a la `_Get_final_iterator_unwrapped` but iterating from the "middle" iterator instead of from the beginning of the range when it falls back to the "increment over and over until we find the end" case.

Reply:
I went with converting `_Last` through `_Get_final_iterator_unwrapped` in the `operator()` and only working on iterators in the actual implementation.

Will push once this is merged